web/recv.js

```diff
@@ -0,0 +1,401 @@
+          height: { min: 720, ideal: 1080 },
+          aspectRatio: matchMedia('all and (orientation:landscape)').matches ? 16 / 9 : 9 / 16,
+          facingMode: 'environment',
+          exposureMode: 'continuous',
+          focusMode: 'continuous',
+          frameRate: { ideal: 15, max: 30 }, // we're not trying to set the user's phone on fire
+        }
+      };
+
+      if (!navigator.mediaDevices || !navigator.mediaDevices.getUserMedia) {
+        return Recv.set_error('mediaDevices not supported? :(');
+      }
+
+      navigator.mediaDevices.getUserMedia(constraints)
+        .then(localMediaStream => {
+          //console.log(localMediaStream);
+          //console.dir(video);
+          if ('srcObject' in video) {
+            video.srcObject = localMediaStream;
+          } else {
+            video.src = URL.createObjectURL(localMediaStream); //deprecated
+          }
+          video.play();
+          video.requestVideoFrameCallback(Recv.on_frame);
+        })
+        .catch(err => {
+          console.error(`OH NO!!!!`, err);
+        });
+    },
+
+    restart_video: function () {
+      console.log("restart video?");
+      if (_video) {
+        return;
+      }
+      if (!('srcObject' in _video)) {
+        console.log("can't restart video? :|")
+        return;
+      }
+      // ios cancels the feed sometimes. Restart it...?
+      // check if localMediaStream.muted after a while???
+      if (_video.srcObject.muted) {
+        Recv.init_video(_video);
+      }
+    },
+
+    download_bytes: function (buff, name) {
+      var blob = new Blob([buff], { type: 'application/octet-stream' });
+      var bloburl = window.URL.createObjectURL(blob);
+      _downloadHelper(name, bloburl);
+      setTimeout(function () {
+        return window.URL.revokeObjectURL(bloburl);
+      }, 1000);
+    },
+
+    on_decode: function (wid, data) {
+      console.log('Main thread received message from worker' + wid + ':', data);
+      // if extract but no bytes, log extract counter
+      if (data.nodata) {
+        _recentExtract = _counter;
+      }
+      if (data.res) {
+        Recv.set_HTML("t" + wid, "avg red " + wid + " is " + data.res);
+        return;
+      }
+      if (data.ready) {
+        if (_workerReady)
+          _workerReady();
+        return;
+      }
+
+      // should be a decode with some bytes, so set decodecounter
+      _recentDecode = _counter;
+
+      const buff = new Uint8Array(data);
+      Recv.set_HTML("t" + wid, "len() is " + buff.length + ", buff: " + buff);
+      Sink.on_decode(buff);
+    },
+
+    on_frame: function (now, metadata) {
+      //console.log("on frame");
+      // https://developer.mozilla.org/en-US/docs/Web/API/VideoFrame
+
+      _counter += 1;
+      if (_workers.length == 0)
+        return;
+      if (_nextWorker >= _workers.length)
+        _nextWorker = 0;
+
+      // piggyback off this call to make sure our visual state is correct
+      Recv.update_visual_state();
```

Comment:
we're calling this here because this is the function that runs on an interval. If we wanted to, we could do setInterval(update_visual_state) and not worry about it here. :thinking: